luaa.c

```diff
@@ -411,6 +412,126 @@ get_modifier_name(int map_index)
+    // every value (Unicode is defined around bytes)
+    if (length < 1 || length > 4)
+        return 0;
+    if (length == 1)
+    {
+        if ((*input & 0x80) == 0)
+            return *input;
+        else
+            return 0;
+    }
+    uint32_t mask = (0xff << (7 - length)) & 0x000000ff;
+    const uint32_t check = (mask << 1) & 0x000000ff;
+    uint32_t ucs = 0x00000000;
+    if ((*input & mask) == check)
+    {
+        mask = (~mask) & 0x000000ff;
+        ucs = (*input & mask) << (6 * (length-1));
+    }
+    else
+        return 0;
+    for (size_t i = 1; i < length; i++)
+    {
+        if ((*(input + i) & 0x000000c0) != 0x00000080)
+            return 0;
+        ucs += (*(input + i) & 0x0000003f) << (6 * (length-i-1));
+    }
+    return ucs;
+}
+
+/** Get Xlib keysym and xkb_keysym_to_uf8 from an Awesome keycode.
+ *
+ * xkb_keysym_to_utf8 returns a UTF-8 representation of the symbol a given
+ * symkey generates. It matches the engraving of the key for level-0 symbols
+ * (but lowercase)
+ *
+ * Keycodes may be given in a string in any valid format for `awful.key`:
+ * "#" + keycode, the symkey name and the UTF-8 representation will all work.
+ *
+ * If no suitable keysym is found, or a malformed keycode is given as an
+ * argument, this function will return (nil, nil)
+ *
+ * @treturn[1] string keysym The keysym name
+ * @treturn[1] nil keysym If no valid keysym is found
+ * @treturn[2] string printsymbol The xkb_keysym_to_utf8 result
+ * @treturn[2] nil printsymbol If the keysym has no printable representation.
+ * @staticfct get_key_name
+ */
+
+static int
+luaA_get_key_name(lua_State *L)
+{
+    // check if argument is valid
+    if (lua_gettop(L) > 1 || lua_type(L, 1) != LUA_TSTRING)
+    {
+        return 0;
+    }
+
+    const char* input = luaL_checkstring(L, 1);
+    const xkb_keysym_t *keysyms;
+    xkb_keysym_t keysym = XKB_KEY_NoSymbol;
+    const size_t length = strlen(input);
+    uint32_t ucs;
+
+    /* Checking for the three possible syntaxes awful.key uses:
+     * 1: #keycode (#8 to #255, any other is invalid)
+     * 2: the symbol itself (the result of xkb_keysym_to_utf8, e.g. @ for at).
+     * 3: the keysym
+     */
+    if (length > 1 && *input == '#' && length < 5) // syntax #1
+    {
+        int keycode_from_hash = 0;
+        for (size_t i = 1; i < length; i++)
+        {
+            if ('9' < *(input + i) || *(input + i) < '0')
+                return 0;
+            else
+            {
+                keycode_from_hash *= 10;
+                keycode_from_hash += input[i] - 48;
+            }
+        }
+        // A keycode cannot go lower than 8 or higher than 255
+        if ((keycode_from_hash < 8) || keycode_from_hash > 255)
+            return 0;
+        xkb_keycode_t keycode = (xkb_keycode_t) keycode_from_hash;
+        struct xkb_keymap *keymap = xkb_state_get_keymap(globalconf.xkb_state);
+        xkb_keymap_key_get_syms_by_level(keymap, keycode, 0, 0, &keysyms);
```

Comment:
Why is it okay to hardcode `layout=0` and `level=0` here? I am not actually sure what they do, but I bet eventually someone will have a use for that.
(Note that I do not really understand XKB, so no idea what I am even suggesting here)

Reply:
Line 508 is pure copypaste. I copied it somewhere from `luaa.c`.

I can justify `level=0`: levels represent modifiers like Shift, AltGr, Shift+AltGr... we want the base keysym with no modifiers because that's what's printed on the key. The modifiers are going to be printed separately in the cheatsheet anyways: (e.g. `Shift+e`, not `E`).

I'm not sure about `layout=0`, though.

Reply:
Well, okay, then I guess this is fine. I bet I already complained at whichever patch added that hardcoded `level=0`.

(So... with my `de` keyboard layout, there is a key with `#`, but no key with `'` because that's Shift+#.... Some day one might want an API that can handle "more".)